include/ginkgo/extensions/kokkos/spaces.hpp

```diff
@@ -0,0 +1,290 @@
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const HipExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const DpcppExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    GKO_NOT_IMPLEMENTED;
+}
+
+
+/**
+ * Throws if the memory space is ~not~ accessible by the executor associated
+ * with the passed in Ginkgo object.
+ *
+ * @tparam MemorySpace  A Kokkos memory space type.
+ * @tparam T  A Ginkgo type that has the member function `get_executor`.
+ *
+ * @param obj  Object which executor is used to check the access to the memory
+ * space.
+ * @param space  The Kokkos memory space to compare agains.
+ */
+template <typename MemorySpace, typename T>
+inline void assert_compatibility(T&& obj, MemorySpace space)
+{
+    if (!check_compatibility(space, obj.get_executor())) {
+        throw Error(__FILE__, __LINE__,
+                    "Executor type and memory space are incompatible");
+    }
+}
+
+
+}  // namespace detail
+
+
+/**
+ * Creates an Executor matching the Kokkos::DefaultHostExecutionSpace.
+ *
+ * If no kokkos host execution space is enabled, this throws an exception.
+ *
+ * @return  An executor of type either ReferenceExecutor or OmpExecutor.
+ */
+inline std::shared_ptr<Executor> create_default_host_executor()
+{
+    static std::mutex mutex{};
+    std::lock_guard<std::mutex> guard(mutex);
+#ifdef KOKKOS_ENABLE_SERIAL
+    if constexpr (std::is_same_v<Kokkos::DefaultHostExecutionSpace,
+                                 Kokkos::Serial>) {
+        static auto exec = ReferenceExecutor::create();
+        return exec;
+    }
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+    if constexpr (std::is_same_v<Kokkos::DefaultHostExecutionSpace,
+                                 Kokkos::OpenMP>) {
+        static auto exec = OmpExecutor::create();
+        return exec;
+    }
+#endif
+    GKO_NOT_IMPLEMENTED;
+}
+
+
+/**
+ * Creates an Executor for a specific Kokkos ExecutionSpace.
+ *
+ * This function supports the following Kokkos ExecutionSpaces:
+ * - Serial
+ * - OpenMP
+ * - Cuda
+ * - HIP
+ * - Experimental::SYCL
+ * If none of these spaces are enabled, then this function throws an exception.
+ * For Cuda, HIP, SYCL, the device-id used by Kokkos is passed to the Executor
+ * constructor.
+ *
+ * @tparam ExecSpace  A supported Kokkos ExecutionSpace.
+ *
+ * @return  An executor matching the type of the ExecSpace.
+ */
+template <typename ExecSpace,
+          typename MemorySpace = typename ExecSpace::memory_space>
+inline std::shared_ptr<Executor> create_executor(ExecSpace ex, MemorySpace = {})
+{
+    static_assert(
+        Kokkos::SpaceAccessibility<ExecSpace, MemorySpace>::accessible);
+    static std::mutex mutex{};
+    std::lock_guard<std::mutex> guard(mutex);
+#ifdef KOKKOS_ENABLE_SERIAL
+    if constexpr (std::is_same_v<ExecSpace, Kokkos::Serial>) {
+        return ReferenceExecutor::create();
```

Comment:
previous function to create host executor will always return the same one. should it be the same here? (not for the device, I guess)

Reply:
That was some left over. Now, the functions will always return different executors, and the user must store them properly.